File: TidbitsTrivia/Core/Engine/SeededRNG.swift
import Foundation

/// Deterministic RNG so the Daily puzzle is identical for everyone on a
/// given day, and so distractor selection is reproducible. Splitmix64.
nonisolated struct SeededRNG: RandomNumberGenerator {
    private var state: UInt64
    init(seed: UInt64) { state = seed &+ 0x9E3779B97F4A7C15 }
    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }
}

extension String {
    /// Stable 64-bit hash (FNV-1a) — used to seed daily/topic generation
    /// reproducibly across launches (Swift's Hasher is per-run salted).
    var stableSeed: UInt64 {
        var hash: UInt64 = 0xCBF29CE484222325
        for byte in utf8 { hash = (hash ^ UInt64(byte)) &* 0x100000001B3 }
        return hash
    }
}

File: windows/Tidbits.Core/Engine/SeededRng.cs
using System.Text;

namespace Tidbits.Core.Engine;

/// Deterministic splitmix64 RNG (port of Core/Engine/SeededRNG.swift). `unchecked`
/// + `ulong` mirrors Swift's wrapping `&+`/`&*` exactly, so distractor selection
/// and any seeded shuffle reproduce bit-for-bit across platforms.
public struct SeededRng
{
    private ulong _state;

    public SeededRng(ulong seed)
    {
        unchecked { _state = seed + 0x9E3779B97F4A7C15UL; }
    }

    public ulong Next()
    {
        unchecked
        {
            _state += 0x9E3779B97F4A7C15UL;
            var z = _state;
            z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9UL;
            z = (z ^ (z >> 27)) * 0x94D049BB133111EBUL;
            return z ^ (z >> 31);
        }
    }
}

public static class StableSeed
{
    /// FNV-1a 64-bit over UTF-8 bytes — reproducible across launches/platforms
    /// (unlike a per-run-salted hash). Byte-exact with the Swift/Kotlin/JS twins.
    public static ulong Of(string s)
    {
        unchecked
        {
            ulong hash = 0xCBF29CE484222325UL;
            foreach (var b in Encoding.UTF8.GetBytes(s))
                hash = (hash ^ b) * 0x100000001B3UL;
            return hash;
        }
    }
}

/// Compares strings by their UTF-8 byte sequences (Swift's
/// `String.utf8.lexicographicallyPrecedes`) — NOT UTF-16 code units, so it stays
/// byte-exact for non-ASCII. Used for platform-agnostic tie-breaks.
public sealed class Utf8Ordinal : IComparer<string>
{
    public static readonly Utf8Ordinal Instance = new();

    public int Compare(string? x, string? y)
    {
        var bx = Encoding.UTF8.GetBytes(x ?? "");
        var by = Encoding.UTF8.GetBytes(y ?? "");
        int n = Math.Min(bx.Length, by.Length);
        for (int i = 0; i < n; i++)
            if (bx[i] != by[i]) return bx[i].CompareTo(by[i]); // byte = unsigned, matches Swift UInt8
        return bx.Length.CompareTo(by.Length);
    }
}
